marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/BuildConstraintsSpec.hs

```diff
@@ -223,3 +248,99 @@ withdrawSpec = Hspec.describe "buildWithdrawConstraints" do
+          :* K (oneSecondEraSummary 4) -- Alonzo lasted 1 second
+          :* K (unboundedEraSummary 5) -- Babbage never ends
+          :* Nil
+      genTipSlot = chooseInteger (9, 20) -- Make sure the tip is in the Babbage era.
+      genMinTime = oneof
+                   [
+                     (1000 *) <$> chooseInteger (6, 20)  -- Even seconds, so there is a chance of collision.
+                   , chooseInteger (6000, 20000)         -- Milliseconds, so there is rounding off.
+                   ]
+      genTimeout = oneof
+                   [
+                     (1000 *) <$> chooseInteger (6, 20)  -- Even seconds, so there is a chance of collision.
+                   , chooseInteger (6000, 20000)         -- Milliseconds, so there is rounding off.
+                   ]
+    Hspec.QuickCheck.prop "valid slot interval for timed-out contract" \assets utxo address marloweParams-> do
+      -- The choice intervals for the tip, minimum time, and timeout overlap, so every ordering will occur.
+      tipSlot' <- genTipSlot
+      minTime <- genMinTime
+      timeout <- genTimeout
+      timeout' <- genTimeout
+      let
+        contract = Semantics.When [] (POSIXTime timeout) Semantics.Close
+        contract' = Semantics.When [] (POSIXTime timeout) $ Semantics.When [] (POSIXTime timeout') Semantics.Close
+      marloweContract <- elements [contract, contract'] -- This contract can only time out.
+      let
+        toSlot = (`div` 1000)
+        tipSlot = Chain.SlotNo $ fromInteger tipSlot'
+        tipTime = 1000 * tipSlot'
+        marloweState = Semantics.State AM.empty AM.empty AM.empty $ POSIXTime minTime
+        datum = Semantics.MarloweData{..}
+        marloweOutput = TransactionScriptOutput{..}
+        result =
+          buildApplyInputsConstraints
+            systemStart eraHistory MarloweV1
+            marloweOutput
+            tipSlot
+            (Chain.TransactionMetadata mempty) Nothing Nothing mempty
+      pure
+        . counterexample ("tipTime = " <> show tipTime)
+        . counterexample ("minTime = " <> show minTime)
+        . counterexample ("contract = " <> show marloweContract)
+        . counterexample ("result = " <> show result)
+        $ case result of
+            Right _ ->
+              counterexample "A valid transaction will occur if tip is not before the first timeout."
+                $ toSlot timeout <= tipSlot'
+            Left (ApplyInputsConstraintsBuildupFailed (MarloweComputeTransactionFailed "TEUselessTransaction")) ->
+              counterexample "A useless transaction will occur if the tip is before the timeout."
+                $ tipTime < timeout
+            Left (ApplyInputsConstraintsBuildupFailed (MarloweComputeTransactionFailed message)) ->
+              if "TEIntervalError (IntervalInPastError " `isPrefixOf` message
+                then counterexample "If the tip is in the past, then the interval should be in the past."
```

Comment:
I think the statement in the counter example is not what we are testing but the converse, same for the next branch

Reply:
I believe it's an "if and only if", but the comment should read the same as the logic.

Reply:
I reworded so the statement matches the logic.

Reply:
My thinking was that the case and nested if basically checks that the tip is in the past only if there is an error. So the tip could be in the past and, if there isn't an error, it wouldn't be checked that it isn't in the past, I think

Reply:
Maybe I am misunderstanding